lib/uplink/main.go

```diff
@@ -0,0 +1,238 @@
+	MaxInlineSize int
+	SegmentSize   int64
+}
+
+// Uplink represents the main entrypoint to Storj V3. An Uplink connects to
+// a specific Satellite and caches connections and resources, allowing one to
+// create sessions delineated by specific access controls.
+type Uplink struct {
+}
+
+// NewUplink creates a new Uplink
+func NewUplink(identity Identity, satelliteAddr string, cfg Config) *Uplink {
+	panic("TODO")
+}
+
+// A Share is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+type Share struct {
+	Access Macaroon
+
+	// TODO: these should be per-bucket somehow maybe? oh man what a nightmare
+	PathCipher       storj.Cipher
+	EncPathPrefix    storj.Path
+	Key              storj.Key
+	EncryptionScheme storj.EncryptionScheme
+}
+
+// ParseShare parses a serialized Share
+func ParseShare(data []byte) (Share, error) {
+	panic("TODO")
+}
+
+func (s *Share) Serialize() ([]byte, error) {
+	panic("TODO")
+}
+
+// Session represents a specific access session.
+type Session struct {
+}
+
+// A Session is created with a Share.
+func (u *Uplink) Session(share Share) *Session {
+	panic("TODO")
+}
+
+// GetBucket returns info about the requested bucket if authorized
+func (s *Session) GetBucket(ctx context.Context, bucket string) (storj.Bucket,
+	error) {
+	panic("TODO")
+}
+
+type CreateBucketOptions struct {
+	PathCipher Cipher
+	// this differs from storj.CreateBucket's choice of just using storj.Bucket
+	// by not having 2/3 unsettable fields.
+}
+
+// CreateBucket creates a new bucket if authorized
+func (s *Session) CreateBucket(ctx context.Context, bucket string,
+	opts *CreateBucketOptions) (storj.Bucket, error) {
+	panic("TODO")
+}
+
+// DeleteBucket deletes a bucket if authorized
+func (s *Session) DeleteBucket(ctx context.Context, bucket string) error {
+	panic("TODO")
+}
+
+// ListBuckets will list authorized buckets
+func (s *Session) ListBuckets(ctx context.Context, opts storj.BucketListOptions) (
+	storj.BucketList, error) {
+	panic("TODO")
+}
+
+// Share creates a new share, potentially further restricted from the Share used
+// to create this session.
+func (s *Session) Share(ctx context.Context, caveats ...Caveat) (Share, error) {
+	panic("TODO")
+}
+
+// ObjectMeta represents metadata about a specific Object
+type ObjectMeta struct {
+	Bucket   string
+	Path     storj.Path
+	IsPrefix bool
+
+	Metadata map[string]string
+
+	Created  time.Time
+	Modified time.Time
+	Expires  time.Time
+
+	Size     int64
+	Checksum string
+
+	// this differs from storj.Object by not having Version (yet), and not
+	// having a Stream embedded. I'm also not sold on splitting ContentType out
+	// from Metadata but shrugemoji.
```

Comment:
Ah, `ContentType` and the `storj.Object` etc. aren't yet completely finalized.

Reply:
oh, cool